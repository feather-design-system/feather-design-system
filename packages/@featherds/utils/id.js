const getSafeId = function (str) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
