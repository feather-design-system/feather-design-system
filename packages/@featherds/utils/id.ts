const getSafeId = function (str: string) {
  str = str || "feather";
  const random = Math.floor(Math.random() * 1000000000);
  return [str.replace(/\s+/g, "-"), Date.now(), random].join("-");
};

export { getSafeId };
