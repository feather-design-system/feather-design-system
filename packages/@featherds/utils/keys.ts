const KEYCODES = {
  PAGEUP: 33,
  PAGEDOWN: 34,
  UP: 38,
  DOWN: 40,
  LEFT: 37,
  RIGHT: 39,
  HOME: 36,
  END: 35,
  ENTER: 13,
  SPACE: 32,
  ESCAPE: 27,
  BACKSPACE: 8,
  DELETE: 46,
};

export { KEYCODES };
