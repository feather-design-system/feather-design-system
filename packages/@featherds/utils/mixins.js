var translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (key) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};
export { translateKey };
