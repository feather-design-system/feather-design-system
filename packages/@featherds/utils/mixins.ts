const translateKey = {
  props: {
    translationKey: {
      type: String,
      default: "",
    },
  },
  methods: {
    translate: function (this: IComponent, key: string) {
      return this.$t
        ? this.$t(this.translationKey + key)
        : "Vue i18n is not loaded";
    },
  },
};

interface IComponent {
  translationKey: string;
  $t: (s: string) => void;
}
export { translateKey };
