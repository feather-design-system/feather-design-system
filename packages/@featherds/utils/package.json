{
  "name": "@featherds/utils",
  "version": "0.12.31",
  "publishConfig": {
    "access": "public"
  },
  "author": "NantHealth",
  "scripts": {
    "compile": "tsc --declaration"
  },
  "keywords": [
    "vue",
    "FeatherUtils"
  ]
}
