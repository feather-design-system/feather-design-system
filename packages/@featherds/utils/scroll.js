var toView = function (el, parent) {
  if (!el || !parent) {
    return;
  }
  var elHeight = el.getBoundingClientRect().height;
  const compStyle = getComputedStyle(el);
  elHeight += parseInt(compStyle["margin-top"], 10);
  elHeight += parseInt(compStyle["margin-bottom"], 10);
  parent.scrollTop =
    el.offsetTop - parent.getBoundingClientRect().height + elHeight;
};

export { toView };
