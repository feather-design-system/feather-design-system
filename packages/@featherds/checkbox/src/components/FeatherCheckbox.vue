<template>
  <div class="layout-container">
    <div
      class="feather-checkbox feather-form-input feather-checkbox-table"
      @click="click"
      @keydown="keydown"
      :aria-checked="modelValue ? 'true' : indeterminate ? 'mixed' : 'false'"
      :aria-disabled="disabled"
      :aria-labelledby="labelId"
      :aria-label="label"
      :id="inputId"
      :tabindex="noFocus ? -1 : 0"
      ref="input"
      role="checkbox"
      data-ref-id="feather-checkbox"
    >
      <div class="checkbox hover focus">
        <div class="box">
          <svg
            class="checkmark"
            role="presentation"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              class="path"
              fill="none"
              d="M1.73,12.91 8.1,19.28 22.79,4.59"
            />
          </svg>
          <div class="indeterminate"></div>
        </div>
        <feather-ripple center v-if="!disabled" />
      </div>
      <label
        v-if="!label"
        data-ref-id="feather-checkbox-label"
        :id="labelId"
        :for="inputId"
      >
        <slot></slot>
      </label>
    </div>
  </div>
</template>
<script>
import { getSafeId } from "@featherds/utils/id";
import { KEYCODES } from "@featherds/utils/keys";
import { FeatherRipple } from "@featherds/ripple";

export default {
  model: {
    prop: "modelValue",
    event: "update:modelValue",
  },
  emits: ["click", "update:modelValue", "indeterminate"],
  props: {
    disabled: {
      type: Boolean,
      default: false,
    },
    modelValue: {
      type: Boolean,
      default: false,
    },
    indeterminate: {
      type: Boolean,
      default: false,
    },
    label: {
      type: String,
    },
    noFocus: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    inputId() {
      return this.label && this.label.length
        ? undefined
        : getSafeId("checkbox");
    },
    labelId() {
      return this.label && this.label.length
        ? undefined
        : getSafeId("checkbox-label");
    },
  },
  watch: {
    indeterminate: {
      handler(v) {
        this.$emit("update:modelValue", undefined);
        this.$emit("indeterminate", v);
      },
    },
  },
  methods: {
    focus() {
      this.$refs.input.focus();
    },
    updateValue() {
      if (!this.disabled) {
        if (this.indeterminate) {
          this.$emit("update:modelValue", true);
          this.$emit("indeterminate", false);
        } else {
          this.$emit("update:modelValue", !this.modelValue);
          this.$emit("indeterminate", false);
        }
      }
    },
    click(e) {
      this.focus();
      this.updateValue();
      this.$emit("click", e);
    },
    keydown(e) {
      if (e.keyCode === KEYCODES.SPACE || e.keyCode === KEYCODES.ENTER) {
        this.updateValue();
      }
      //stop page scrolling
      if (e.keyCode === KEYCODES.SPACE) {
        e.preventDefault();
      }
    },
  },
  components: {
    FeatherRipple,
  },
};
</script>

<style lang="scss" scoped>
@use "sass:math";
@import "@featherds/styles/themes/variables";
@import "@featherds/styles/themes/utils";
@import "@featherds/styles/mixins/typography";

.feather-checkbox {
  @include state-on-surface();
}
.layout-container {
  display: inline-block;
  margin-bottom: $spacing-xs;
}

$height: 36px;
$boxHeight: math.div($height, 2);
.feather-checkbox {
  display: inline-flex;
  margin-left: -8px;
  cursor: pointer;
  align-items: center;
  .checkbox {
    height: $height;
    width: $height;
    border-radius: 100%;
    position: relative;
    flex: none;
    align-items: center;
    justify-content: center;
    display: flex;
  }
  .box {
    height: $boxHeight;
    width: $boxHeight;
    border: 2px solid var($secondary-text-on-surface);
    background-color: transparent;
    border-radius: 2px;
    pointer-events: none;
    display: inline-flex;
    justify-content: center;
    align-content: center;
    position: relative;
    transition: border-color 90ms 0ms cubic-bezier(0, 0, 0.2, 1),
      background-color 90ms 0ms cubic-bezier(0, 0, 0.2, 1);
  }
  .checkmark {
    opacity: 0;
    color: var($surface);
    flex: 1;
    transition: opacity 180ms 0ms cubic-bezier(0, 0, 0.2, 1),
      transform 180ms 0ms cubic-bezier(0, 0, 0.2, 1);
    .path {
      transition: stroke-dashoffset 280ms 0ms cubic-bezier(0.4, 0, 0.6, 1);
      stroke: currentColor;
      stroke-width: 3.12px;
      stroke-dashoffset: 30;
      stroke-dasharray: 30;
    }
  }
  .indeterminate {
    position: absolute;
    left: 0px;
    border-width: 1px;
    border-style: solid;
    border-bottom-color: var($surface);
    border-top-color: var($surface);
    height: 0;
    width: 100%;
    margin-top: math.div($boxHeight - 6, 2);
    opacity: 0;
    transform: rotateZ(135deg);
    transition: opacity 180ms 0ms cubic-bezier(0, 0, 0.2, 1),
      transform 180ms 0ms cubic-bezier(0, 0, 0.2, 1);
  }
  label {
    @include body-small();
    color: var($primary-text-on-surface);
    padding-left: 4px;
    padding-right: 16px;
    min-width: 56px;
    user-select: none;
    -khtml-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    -o-user-select: none;
    -webkit-user-select: none;
    cursor: pointer;
  }
}
.feather-checkbox:hover[aria-checked="true"] .box {
  border-color: var($primary);
  background-color: var($primary);
}

.feather-checkbox[aria-checked="true"] .box {
  border-color: var($primary);
  background-color: var($primary);
  .checkmark {
    opacity: 1;
  }
  .path {
    stroke-dashoffset: 0;
  }
}

.feather-checkbox[aria-checked="mixed"] .box {
  border-color: var($primary);
  background-color: var($primary);
  .checkmark {
    opacity: 0;
  }
  .indeterminate {
    opacity: 1;
    transform: rotateZ(0deg);
  }
}

.feather-checkbox[aria-disabled="true"] {
  @include state-disabled();
  cursor: default;
  .box {
    border-color: var($disabled-text-on-surface);
  }
  .checkbox {
    background: transparent;
  }
  &[aria-checked="true"] .box {
    border-color: transparent;
    background-color: var($disabled-text-on-surface);
  }
  label {
    color: var($disabled-text-on-surface);
    cursor: default;
  }
}
</style>
