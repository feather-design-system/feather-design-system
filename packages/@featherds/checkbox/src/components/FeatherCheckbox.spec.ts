import { nextTick } from "vue";
import axe from "@featherds/utils/test/axe";
import { mount } from "@vue/test-utils";
import FeatherCheckbox from "./FeatherCheckbox.vue";
import * as id from "@featherds/utils/id";
import { getCalls } from "@featherds/utils/test/calls";
import { Code } from "@featherds/utils/keys";
jest.spyOn(id, "getSafeId").mockImplementation((x) => x);

const getWrapper = function (options: Record<string, unknown> = {}) {
  options = {
    ...options,
    ...getSlot(),
    global: {
      provide: {
        registerCheckbox: jest.fn(),
      },
    },
  };
  return mount(FeatherCheckbox, options);
};
const getSlot = () => ({
  slots: {
    default: {
      template: "<span>Hello</span>",
    },
  },
});

describe("FeatherCheckbox.vue", () => {
  it("should have a label", () => {
    const wrapper = getWrapper();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should set aria-checked based on property", async () => {
    const wrapper = getWrapper({
      props: {
        modelValue: true,
      },
    });
    expect(wrapper.element).toMatchSnapshot();

    wrapper.setProps({
      modelValue: false,
    });
    await nextTick();
    expect(wrapper.element).toMatchSnapshot();
    wrapper.setProps({
      modelValue: false,
      indeterminate: true,
    });
    await nextTick();
    expect(wrapper.element).toMatchSnapshot();
  });
  it("should not emit input event when disabled", async () => {
    const wrapper = getWrapper({
      props: {
        disabled: true,
      },
    });
    await wrapper.find("[role='checkbox']").trigger("click");
    expect(wrapper.emitted("update:modelValue")).toBeFalsy();
  });
  it("should emit input event when clicked", async () => {
    const wrapper = getWrapper();
    await wrapper.find("[role='checkbox']").trigger("click");
    expect(getCalls<boolean[]>(wrapper, "update:modelValue")[0][0]).toBe(true);
  });
  it("should toggle the value when enter or space is pressed", async () => {
    const testKeydown = async (code: Code, modelValue: boolean) => {
      const wrapper = getWrapper({
        props: { modelValue },
      });
      expect(wrapper.vm.modelValue).toBe(modelValue);
      const input = wrapper.find<HTMLElement>("[role='checkbox']");
      await input.trigger("keydown", { code });
      expect(getCalls<boolean[]>(wrapper, "update:modelValue")[0][0]).toBe(
        !modelValue
      );
    };
    await testKeydown(Code.SPACE, false);
    await testKeydown(Code.ENTER, true);
    await testKeydown(Code.ENTER, false);
    await testKeydown(Code.SPACE, true);
  });
  it("should not display a label when label property is used", () => {
    const label = "test";
    const wrapper = getWrapper({
      props: {
        label,
      },
    });
    expect(wrapper.find("label").exists()).toBe(false);
    expect(wrapper.find("[role='checkbox']").attributes("aria-label")).toBe(
      label
    );
  });
  it("should set tabindex to -1 when noFocus is used", () => {
    const wrapper = getWrapper({
      props: {
        noFocus: true,
      },
    });
    expect(wrapper.find("[role='checkbox']").attributes("tabindex")).toBe("-1");
  });
  describe("indeterminate", () => {
    it("should set value to 'undefined' when indeterminate is set to true", async () => {
      const wrapper = getWrapper({
        props: {
          modelValue: true,
        },
      });
      wrapper.setProps({
        indeterminate: true,
      });
      await nextTick();
      expect(
        getCalls<boolean[]>(wrapper, "update:modelValue")[0][0]
      ).toBeUndefined();
    });
    it("should set value to true when an indeterminate checkbox is clicked", async () => {
      const wrapper = getWrapper({
        props: {
          modelValue: true,
        },
      });
      wrapper.setProps({
        indeterminate: true,
      });
      await nextTick();
      await wrapper.find("[role='checkbox']").trigger("click");
      expect(getCalls<boolean[]>(wrapper, "update:modelValue")[1][0]).toBe(
        true
      );
    });
    it("should emit an indeterminate event when indeterminate is changed via click of checkbox", async () => {
      const wrapper = getWrapper({
        props: {
          modelValue: true,
        },
      });
      wrapper.setProps({
        indeterminate: true,
      });
      await nextTick();
      await wrapper.find("[role='checkbox']").trigger("click");
      expect(getCalls<boolean[]>(wrapper, "indeterminate")[1][0]).toBe(false);
    });
    it("should emit an indeterminate event when indeterminate is set via prop", async () => {
      const wrapper = getWrapper({
        props: {
          modelValue: true,
        },
      });
      wrapper.setProps({
        indeterminate: true,
      });
      await nextTick();
      expect(getCalls<boolean[]>(wrapper, "indeterminate")[0][0]).toBe(true);
    });
  });
  describe("a11y", () => {
    const accessibilityTest = async (options: Record<string, unknown>) => {
      const wrapper = getWrapper(options);
      expect(await axe(wrapper.element)).toHaveNoViolations();
    };
    it("should be accessible when unchecked", async () => {
      await accessibilityTest({});
    });
    it("should be accessible when disabled and unchecked", async () => {
      await accessibilityTest({
        props: {
          disabled: true,
        },
      });
    });
    it("should be accessible when disabled and checked ", async () => {
      await accessibilityTest({
        props: {
          disabled: true,
          modelValue: true,
        },
      });
    });
    it("should be accessible when checked ", async () => {
      await accessibilityTest({
        props: {
          modelValue: true,
        },
      });
    });
  });
});
