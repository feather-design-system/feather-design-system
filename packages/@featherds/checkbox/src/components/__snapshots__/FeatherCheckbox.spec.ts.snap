// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatherCheckbox.vue should have a label 1`] = `
<div
  class="layout-container"
>
  <div
    aria-checked="false"
    aria-disabled="false"
    aria-labelledby="checkbox-label"
    class="feather-checkbox feather-form-input feather-checkbox-table"
    data-ref-id="feather-checkbox"
    id="checkbox"
    role="checkbox"
    tabindex="0"
  >
    <div
      class="checkbox hover focus"
      style="overflow: hidden; position: relative;"
    >
      <div
        class="box"
      >
        <svg
          class="checkmark"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            class="path"
            d="M1.73,12.91 8.1,19.28 22.79,4.59"
            fill="none"
          />
        </svg>
        <div
          class="indeterminate"
        />
      </div>
      <!---->
    </div>
    <label
      data-ref-id="feather-checkbox-label"
      for="checkbox"
      id="checkbox-label"
    >
      
      <span>
        Hello
      </span>
      
    </label>
  </div>
</div>
`;

exports[`FeatherCheckbox.vue should set aria-checked based on property 1`] = `
<div
  class="layout-container"
>
  <div
    aria-checked="true"
    aria-disabled="false"
    aria-labelledby="checkbox-label"
    class="feather-checkbox feather-form-input feather-checkbox-table"
    data-ref-id="feather-checkbox"
    id="checkbox"
    role="checkbox"
    tabindex="0"
  >
    <div
      class="checkbox hover focus"
      style="overflow: hidden; position: relative;"
    >
      <div
        class="box"
      >
        <svg
          class="checkmark"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            class="path"
            d="M1.73,12.91 8.1,19.28 22.79,4.59"
            fill="none"
          />
        </svg>
        <div
          class="indeterminate"
        />
      </div>
      <!---->
    </div>
    <label
      data-ref-id="feather-checkbox-label"
      for="checkbox"
      id="checkbox-label"
    >
      
      <span>
        Hello
      </span>
      
    </label>
  </div>
</div>
`;

exports[`FeatherCheckbox.vue should set aria-checked based on property 2`] = `
<div
  class="layout-container"
>
  <div
    aria-checked="false"
    aria-disabled="false"
    aria-labelledby="checkbox-label"
    class="feather-checkbox feather-form-input feather-checkbox-table"
    data-ref-id="feather-checkbox"
    id="checkbox"
    role="checkbox"
    tabindex="0"
  >
    <div
      class="checkbox hover focus"
      style="overflow: hidden; position: relative;"
    >
      <div
        class="box"
      >
        <svg
          class="checkmark"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            class="path"
            d="M1.73,12.91 8.1,19.28 22.79,4.59"
            fill="none"
          />
        </svg>
        <div
          class="indeterminate"
        />
      </div>
      <!---->
    </div>
    <label
      data-ref-id="feather-checkbox-label"
      for="checkbox"
      id="checkbox-label"
    >
      
      <span>
        Hello
      </span>
      
    </label>
  </div>
</div>
`;

exports[`FeatherCheckbox.vue should set aria-checked based on property 3`] = `
<div
  class="layout-container"
>
  <div
    aria-checked="mixed"
    aria-disabled="false"
    aria-labelledby="checkbox-label"
    class="feather-checkbox feather-form-input feather-checkbox-table"
    data-ref-id="feather-checkbox"
    id="checkbox"
    role="checkbox"
    tabindex="0"
  >
    <div
      class="checkbox hover focus"
      style="overflow: hidden; position: relative;"
    >
      <div
        class="box"
      >
        <svg
          class="checkmark"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            class="path"
            d="M1.73,12.91 8.1,19.28 22.79,4.59"
            fill="none"
          />
        </svg>
        <div
          class="indeterminate"
        />
      </div>
      <!---->
    </div>
    <label
      data-ref-id="feather-checkbox-label"
      for="checkbox"
      id="checkbox-label"
    >
      
      <span>
        Hello
      </span>
      
    </label>
  </div>
</div>
`;
